components/gfx/paint_context.rs

```diff
@@ -355,6 +368,102 @@ impl<'a> PaintContext<'a> {
+        // given the equation of a line,
+        // y = a * x + b,
+        // and the equation of an ellipse,
+        // x^2/w^2 + y^2/h^2 = 1,
+        // substitute y = a * x + b, giving
+        // x^2/w^2 + (a^2x^2 + 2abx + b^2)/h^2 = 1
+        // then simplify to
+        // (h^2 + w^2a^2)x^2 + 2abw^2x + (b^2w^2 - w^2h^2) = 0
+        // finally solve for w using the quadratic equation.
+        let w = e.width;
+        let h = e.height;
+        let quad_a = h * h + w * w * a * a;
+        let quad_b = 2. * a * b * w * w;
+        let quad_c = b * b * w * w - w * w * h * h;
+        let intersections = PaintContext::solve_quadratic(quad_a, quad_b, quad_c);
+        match intersections {
+            (Some(x0), Some(x1)) => {
+                let mut p0 = Point2D {x: x0 + e.origin.x, y: a * x0 + b + e.origin.y};
+                let mut p1 = Point2D {x: x1 + e.origin.x, y: a * x1 + b + e.origin.y};
+                if x0 > x1 {
+                    let tmp = p0;
+                    p0 = p1;
+                    p1 = tmp;
+                }
+                (Some(p0), Some(p1))
+            },
+            (Some(x0), None) => {
+                let p = Point2D {x: x0 + e.origin.x, y: a * x0 + b + e.origin.y};
+                (Some(p), None)
+            },
+            (None, Some(x1)) => {
+                let p = Point2D {x: x1 + e.origin.x, y: a * x1 + b + e.origin.y};
+                (Some(p), None)
+            },
+            (None, None) => (None, None),
+        }
+    }
+
+    fn ellipse_line_intersection_angles(e: Ellipse, l: Line)
+                                        -> (Option<(Point2D<f32>, f32)>, Option<(Point2D<f32>, f32)>) {
+        fn point_angle(e: Ellipse, intersect_point: Point2D<f32>) -> f32 {
+            ((intersect_point.y - e.origin.y).abs() / e.height).asin()
+        }
+
+        let intersection = PaintContext::intersect_ellipse_line(e, l);
+        match intersection {
+            (Some(p0), Some(p1)) => (Some((p0, point_angle(e, p0))), Some((p1, point_angle(e, p1)))),
+            (Some(p0), None) => (Some((p0, point_angle(e, p0))), None),
+            (None, Some(p1)) => (None, Some((p1, point_angle(e, p1)))),
+            (None, None) => (None, None),
+        }
+    }
+
+    fn ellipse_rightmost_line_intersection_angle(e: Ellipse, l: Line) -> Option<f32> {
+        match PaintContext::ellipse_line_intersection_angles(e, l) {
+            (Some((p0, angle0)), Some((p1, angle1))) => if p0.x < p1.x { Some(angle1) } else { Some(angle0) },
+            (Some((_, angle0)), None) => Some(angle0),
+            (None, Some((_, angle1))) => Some(angle1),
+            (None, None) => None,
+        }
+    }
+
+    fn ellipse_leftmost_line_intersection_angle(e: Ellipse, l: Line) -> Option<f32> {
+        match PaintContext::ellipse_line_intersection_angles(e, l) {
+            (Some((p0, angle0)), Some((p1, angle1))) => if p0.x < p1.x { Some(angle0) } else { Some(angle1) },
```

Comment:
nit: You could use an `if` guard here in the match to make this line shorter if you like.